### ISSUE TITLE
Detect unused type key

### ISSUE THREAD

**Issue by TomasBarry** [gajus/eslint-plugin-flowtype#407]
Given the following code:

```es6
// @flow

type A = {
  a: number,
  b: number
}

// example one
const f = (a: A): void => console.log(a.a)

// example two
const f = ({ a }: A): void => console.log(a)
```

In either example, is it possible for ESlint to report that the key `b` in type `A` is unused? I would imagine that it would be easier to report this in the second example given that the first example could muddy the waters by allowing for:

```es6
...
// alternative example one
const f = (a: A): void => console.log(a)
```

Whereas example 2 would not allow for the same potential without actually changing the function signature.

Is there a rule that I haven't come across that allows for this type of reporting?

**Comment by pnevyk** [comment]
This is not possible because eslint (and hence this plugin) works just on the syntax tree, so there is basically no knowledge about what type `A` actually is (it would be theoretically possible in your example using a hacky code since the type declaration is in the same file - ie. tree - as the functions, but if the type would come from a different file, a library or even from "standard" library, there is no way to get the type to eslint, at least not yet).

Just for curiosity, do you have a real-world scenario where this rule would be useful? Enforcing the developer to use all properties of a type in a function seems a bit strange to me.

**Comment by TomasBarry** [comment]
It's not that I would like people to use all properties of a type, instead, I would like to be able to ensure that all properties of a type are used somewhere.

Often I find that when an application grows and changes, attributes of a type that once made sense no longer do and having Flow report on that would be preferable to relying on a developer to double check and potentially get it wrong. It's largely for code cleanliness.

I imagine it would be far more difficult to report on that for imported/exported types, but for local types, I'd imagine it would be possible (the same way that if a type is exported, but unused anywhere, Flow does not flag it, but an unused local type declaration that is unused is reported on).

**Comment by pnevyk** [comment]
Sorry, I misunderstood you before, this makes much more sense to me. If I am not overlooking something, it could be possible for local types. But to be safe, the rule would have to handle things like

* type aliases
```js
type B = A;
function g(b: B) { console.log(b.b); }
```

* spread syntax for objects into other types which have the same key which is actually used in them
```js
type B = { b: number };
const a: A = { a: 0, b: 0 };
const b: B = { ... a };
```

and so on. I think that such rule would be quite complicated, not just for implementation, but even for specification.

**Comment by TomasBarry** [comment]
Actually, I have found a case where it is reported on by ESLint, but under specific circumstances. This gives the appropriate warning:

```es6
...
type PresentationalProps = {|
  a: string,
  b: string
|}

type ActionProps = {||}

type Props =
  & PresentationalProps
  & ActionProps

const mapStateToProps = (): PresentationalProps => {
  return { a: 'a', b: 'b' }
}

const component = ({ a }: Props): React.Node => (
  <p>
    { `${a}` }
  </p>
)
...
```

The key feature being the intersection type for `Props`. Although `ActionProps` is an empty object, this will not raise the same warning:

```es6
type PresentationalProps = {|
  a: string,
  b: string
|}

type Props =
  & PresentationalProps
```

The warning being, `react/no-unused-prop-types`. I'm unsure what the logic is that allows for that warning, but if it could be extended to any type usage, and not just in React, even just locally, that would be a really nice feature to have.
